### PR TITLE
Separate vis-specific controls from centralized controls

### PR DESCRIPTION
**Pull request by espressoroaster**
### CATEGORY
- [x] Enhancement (new features, refinement)

### SUMMARY
This PR extends upon @kristw 's proof of concept that allows each visualization to define its own controls, instead of storing all controls in the centralized `controls.jsx` file. Shared controls defined in `controls.jsx` are still supported. Moving forward, we would want to create a registry for controls that are defined at the visualization level. This would allow us to move towards a modularized world where new charts can be installed and maintained as plugins, reducing bloat in our codebase. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
good_behavior

image

### TEST PLAN
Verified that the `rose` chart modified in this PR is working correctly

### ADDITIONAL INFORMATION
- [x] Has associated issue:
This PR uncovered an issue with the `vizType` form data property not being updated when you switch from one visualization to another (ex. Line to Rose, Rose to Pie, etc.). This has likely gone unnoticed because chart controls used to rely on the `controls.jsx` index -- not `vizType`. This new PR requires the correct `vizType` to be passed in so the correct control panel config can be identified and looked up against.

The UI implications of this bug are shown in the GIF below. Note that desired behavior is for `Use Area Proportions` checkbox control to trigger an automatic re-render.

bad_behavior

image

**[Update August 20, 2019]**: SHA de75110 fixes the vizType issue described above.
fixed_renderTrigger

### REVIEWERS
@kristw @williaster

### CODE REVIEW

**Comment by felixcodes** [comment]
Please do not merge this PR until the `vizType` updating issue has been fully looked into and resolved.

File: superset/assets/src/explore/controlUtils.js [hunk 1]
```diff
@@ -44,29 +44,37 @@ export function validateControl(control) {
...
 
 export function getControlConfig(controlKey, vizType) {
   // Gets the control definition, applies overrides, and executes
   // the mapStatetoProps
   const vizConf = controlPanelConfigs[vizType] || {};
   const controlOverrides = vizConf.controlOverrides || {};
-  const control = {
+
+  if (isControlObject(controlKey)) {
+    for (const section of vizConf.controlPanelSections) {
```
Comment (felixcodes) [inline superset/assets/src/explore/controlUtils.js]: TODO: flatmap + filter

**Comment by mistercrunch** [comment]
What is the state of this PR? What's missing to push forward and merge?

**Comment by kristw** [comment]
@mistercrunch There are some issues with initial states that are not propagated properly.

**Comment by felixcodes** [comment]
@kristw Fixed the vizType state issue in SHA 3c641e8. The `renderTrigger` property now works as intended for controls defined at the visualization level.

File: superset/assets/src/explore/controlUtils.js [hunk 2]
```diff
@@ -122,3 +130,22 @@ export function getControlState(controlKey, vizType, state, value) {
   controlState = handleMissingChoice(controlKey, controlState);
   return validateControl(controlState);
 }
+
+export function getAllControlsStateFromFormDataKeys(vizType, datasourceType, state, formData) {
```
Comment (kristw) [inline superset/assets/src/explore/controlUtils.js]: shorten to `getAllControlsState`

File: superset/assets/src/explore/store.js [hunk 3]
```diff
@@ -50,36 +49,41 @@ export function getControlsState(state, inputFormData) {
...
 export function applyDefaultFormData(inputFormData) {
   const datasourceType = inputFormData.datasource.split('__')[1];
   const vizType = inputFormData.viz_type;
-  const controlNames = getControlKeys(vizType, datasourceType);
+  const controlsState =
+    getAllControlsStateFromFormDataKeys(
+      vizType,
+      datasourceType,
+      null,
+      Object.assign({}, inputFormData),
```
Comment (kristw) [inline superset/assets/src/explore/store.js]: replace `Object.assign` with `{...inputFormData}`

**Review by kristw — APPROVED** [review]
few more nits

**Pull request merged**